sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt

```diff
@@ -0,0 +1,225 @@
+    asyncInflater: AsyncLayoutInflater,
+    asyncInflateCallback: (View) -> Unit
+  ) {
+    validateOptions(options)
+    asyncInflater.inflate(resId, mapView) { view, _, _ ->
+      asyncInflateCallback.invoke(prepareViewAnnotation(view, options))
+    }
+  }
+
+  override fun addViewAnnotation(
+    @LayoutRes resId: Int,
+    options: ViewAnnotationOptions
+  ): View {
+    validateOptions(options)
+    val view = LayoutInflater.from(mapView.context).inflate(resId, mapView, false)
+    return prepareViewAnnotation(view, options)
+  }
+
+  override fun addViewAnnotation(view: View, options: ViewAnnotationOptions) {
+    if (idLookupMap.contains(view)) {
+      throw RuntimeException(
+        "Trying to add view annotation that was already added before! " +
+          "Please consider deleting annotation view ($view) beforehand."
+      )
+    }
+    validateOptions(options)
+    prepareViewAnnotation(view, options)
+  }
+
+  override fun removeViewAnnotation(view: View): Boolean {
+    val id = idLookupMap[view] ?: return false
+    val annotation = annotationMap.remove(id) ?: return false
+    idLookupMap.remove(view)
+    view.viewTreeObserver.removeOnGlobalLayoutListener(annotation.globalLayoutListener)
+    mapView.removeView(view)
+    mapboxMap.removeViewAnnotation(id)
+    return true
+  }
+
+  override fun updateViewAnnotation(
+    view: View,
+    options: ViewAnnotationOptions,
+  ): Boolean {
+    val id = idLookupMap[view] ?: return false
+    annotationMap[id]?.let {
+      it.handleVisibilityAutomatically = options.visible == null
+      mapboxMap.updateViewAnnotation(id, options)
+      return true
+    } ?: return false
+  }
+
+  override fun getViewAnnotationByFeatureId(featureId: String): View? {
+    val (view, _) = findByFeatureId(featureId)
+    return view
+  }
+
+  override fun getViewAnnotationOptionsByFeatureId(featureId: String): ViewAnnotationOptions? {
+    val (_, options) = findByFeatureId(featureId)
+    return options
+  }
+
+  override fun getViewAnnotationOptionsByView(view: View): ViewAnnotationOptions? {
+    val id = idLookupMap[view] ?: return null
+    val options = mapboxMap.getViewAnnotationOptions(id)
+    if (options.isValue) {
+      return options.value
+    }
+    return null
+  }
+
+  override fun onViewAnnotationPositionsUpdate(positions: MutableList<ViewAnnotationPositionDescriptor>) {
+    drawAnnotationViews(positions)
+  }
+
+  fun destroy() {
+    mapboxMap.setViewAnnotationPositionsUpdateListener(null)
+    annotationMap.forEach { (id, annotation) ->
+      mapboxMap.removeViewAnnotation(id)
+      annotation.view.viewTreeObserver.removeOnGlobalLayoutListener(annotation.globalLayoutListener)
+      mapView.removeView(annotation.view)
+    }
+    currentViewsDrawnMap.clear()
+    annotationMap.clear()
+    idLookupMap.clear()
+  }
+
+  private fun validateOptions(options: ViewAnnotationOptions) {
+    checkNotNull(options.geometry) { "Geometry can not be null!" }
```

Comment:
Should we maybe check that the geometry is a point?

Reply:
Discussed on Slack that the error is handled on GL-Native side and we have to propagate it through our APIs.

Reply:
tracking in https://github.com/mapbox/mapbox-maps-android/pull/850